Mathlib/RingTheory/PiTensorProduct.lean

```diff
@@ -0,0 +1,248 @@
+  | smul_tprod => simp
+  | add _ _ h1 h2 => simp [h1, h2]
+
+instance instNonAssocSemiring : NonAssocSemiring (⨂[R] i, A i) where
+  __ := instNonUnitalNonAssocSemiring
+  one_mul := PiTensorProduct.one_mul
+  mul_one := PiTensorProduct.mul_one
+
+end NonAssocSemiring
+
+noncomputable section NonUnitalSemiring
+
+variable [CommSemiring R] [∀ i, NonUnitalSemiring (A i)]
+variable [∀ i, Module R (A i)] [∀ i, SMulCommClass R (A i) (A i)] [∀ i, IsScalarTower R (A i) (A i)]
+
+protected lemma mul_assoc (x y z : ⨂[R] i, A i) : mul (mul x y) z = mul x (mul y z) := by
+  -- restate as an equality of morphisms so that we can use `ext`
+  suffices LinearMap.llcomp R _ _ _ mul ∘ₗ mul =
+      (LinearMap.llcomp R _ _ _ LinearMap.lflip <| LinearMap.llcomp R _ _ _ mul.flip ∘ₗ mul).flip by
+    exact DFunLike.congr_fun (DFunLike.congr_fun (DFunLike.congr_fun this x) y) z
+  ext x y z
+  dsimp [← mul_def]
+  simpa only [tprod_mul_tprod] using congr_arg (tprod R) (mul_assoc x y z)
+
+instance instNonUnitalSemiring : NonUnitalSemiring (⨂[R] i, A i) where
+  __ := instNonUnitalNonAssocSemiring
+  mul_assoc := PiTensorProduct.mul_assoc
+
+end NonUnitalSemiring
+
+noncomputable section Semiring
+
+variable [CommSemiring R'] [CommSemiring R] [∀ i, Semiring (A i)]
+variable [Algebra R' R] [∀ i, Algebra R (A i)] [∀ i, Algebra R' (A i)]
+variable [∀ i, IsScalarTower R' R (A i)]
+
+instance instSemiring : Semiring (⨂[R] i, A i) where
+  __ := instNonUnitalSemiring
+  __ := instNonAssocSemiring
+
+instance instAlgebra : Algebra R' (⨂[R] i, A i) where
+  __ := hasSMul'
+  toFun := (· • 1)
+  map_one' := by simp
+  map_mul' r s := show (r * s) • 1 = mul (r • 1) (s • 1)  by
+    rw [LinearMap.map_smul_of_tower, LinearMap.map_smul_of_tower, LinearMap.smul_apply, mul_comm,
+      mul_smul]
+    congr
+    show (1 : ⨂[R] i, A i) = 1 * 1
+    rw [mul_one]
+  map_zero' := by simp
+  map_add' := by simp [add_smul]
+  commutes' r x := by
+    simp only [RingHom.coe_mk, MonoidHom.coe_mk, OneHom.coe_mk]
+    change mul _ _ = mul _ _
+    rw [LinearMap.map_smul_of_tower, LinearMap.map_smul_of_tower, LinearMap.smul_apply]
+    change r • (1 * x) = r • (x * 1)
+    rw [mul_one, one_mul]
+  smul_def' r x := by
+    simp only [RingHom.coe_mk, MonoidHom.coe_mk, OneHom.coe_mk]
+    change _ = mul _ _
+    rw [LinearMap.map_smul_of_tower, LinearMap.smul_apply]
+    change _ = r • (1 * x)
+    rw [one_mul]
+
+lemma algebraMap_apply (r : R') (i : ι) [DecidableEq ι] :
+    algebraMap R' (⨂[R] i, A i) r = tprod R (Pi.mulSingle i (algebraMap R' (A i) r)) := by
+  change r • tprod R 1 = _
+  have : Pi.mulSingle i (algebraMap R' (A i) r) = update (fun i ↦ 1) i (r • 1) := by
+    rw [Algebra.algebraMap_eq_smul_one]; rfl
+  rw [this, ← smul_one_smul R r (1 : A i), MultilinearMap.map_smul, update_eq_self, smul_one_smul]
+  congr
+
+/--
+The map `Aᵢ ⟶ ⨂ᵢ Aᵢ` given by `a ↦ 1 ⊗ ... ⊗ a ⊗ 1 ⊗ ...`
+-/
+@[simps]
+def fromComponentAlgHom [DecidableEq ι] (i : ι) : A i →ₐ[R] ⨂[R] i, A i where
```

Comment:
```suggestion
def tprodSingleAlgHom [DecidableEq ι] (i : ι) : A i →ₐ[R] ⨂[R] i, A i where
```
or perhaps just
```suggestion
def singleAlgHom [DecidableEq ι] (i : ι) : A i →ₐ[R] ⨂[R] i, A i where
```
since the `inl` and `inr` versions for `TensorProduct` are quite succinctly named